sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -1077,23 +1077,37 @@ def set_lpmode(self, lpmode):
         if self.is_flat_memory() or not self.get_lpmode_support():
             return False
 
+        DELAY_RETRY = 0.1
         lpmode_val = self.xcvr_eeprom.read(consts.MODULE_LEVEL_CONTROL)
         if lpmode_val is not None:
             if lpmode is True:
                 # Force module transition to LowPwr under SW control
                 lpmode_val = lpmode_val | (1 << CmisApi.LowPwrRequestSW)
                 self.xcvr_eeprom.write(consts.MODULE_LEVEL_CONTROL, lpmode_val)
-                time.sleep(0.1)
+                start_time = time.time()
+                pwroff_duration = self.get_module_pwr_down_duration()/1000
+                # Loop until the power-off duration has elapsed
+                while (time.time() - start_time) < pwroff_duration:
+                    if self.get_lpmode():
+                        return self.get_lpmode()
+                    # Sleep for a short interval before the next check
+                    time.sleep(DELAY_RETRY)
                 return self.get_lpmode()
             else:
                 # Force transition from LowPwr to HighPower state under SW control.
                 # This will transition LowPwrS signal to False. (see Table 6-12 CMIS v5.0)
                 lpmode_val = lpmode_val & ~(1 << CmisApi.LowPwrRequestSW)
                 lpmode_val = lpmode_val & ~(1 << CmisApi.LowPwrAllowRequestHW)
                 self.xcvr_eeprom.write(consts.MODULE_LEVEL_CONTROL, lpmode_val)
-                time.sleep(1)
-                mstate = self.get_module_state()
-                return True if mstate == 'ModuleReady' else False
+                start_time = time.time()
+                # Loop until the power-on duration has elapsed
+                pwron_duration = self.get_module_pwr_up_duration()/1000
+                while (time.time() - start_time) < pwron_duration:
+                    mstate = self.get_module_state()
+                    if mstate == 'ModuleReady':
+                        return True
+                    # Sleep for a short interval before the next check
+                    time.sleep(DELAY_RETRY)
         return False
```

Comment:
@AnoopKamath We need to check for mstate here as well since the current implementation ignores the module state if the `ModuleReady` transition happens in the last iteration of polling (precisely, after the 0.1s sleep is executed as part of last poll).
This scenario should be handled by the implementation I shared earlier though.

Reply:
Yes. taken care